File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/two/Long_ScreenTwo_EMA_TwoBarsAscendingAndCrossing.java
package com.kishlaly.ta.analyze.tasks.blocks.two;

import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.QuoteJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.EMAJava;
import com.kishlaly.ta.model.indicators.IndicatorJava;
import com.kishlaly.ta.utils.CollectionUtilsJava;
import com.kishlaly.ta.utils.LogJava;

import java.util.List;

import static com.kishlaly.ta.model.BlockResultCodeJava.*;
import static com.kishlaly.ta.utils.Quotes.*;

/**
 * Quotes should cross EMA13 and should go up
 */
public class Long_ScreenTwo_EMA_TwoBarsAscendingAndCrossing implements ScreenTwoBlock {
    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        List<QuoteJava> screen_2_Quotes = screen.quotes;
        List<EMAJava> screen_2_EMA13 = (List<EMAJava>) screen.indicators.get(IndicatorJava.EMA13);
        // prerequisite 1:
        // make sure first that the last TWO columns go up
        QuoteJava preLastQuote = CollectionUtilsJava.getFromEnd(screen_2_Quotes, 2);
        QuoteJava lastQuote = CollectionUtilsJava.getFromEnd(screen_2_Quotes, 1);
        boolean ascendingBarHigh = preLastQuote.getHigh() < lastQuote.getHigh();
        if (!ascendingBarHigh) {
            LogJava.recordCode(QUOTE_HIGH_NOT_GROWING_SCREEN_2, screen);
            LogJava.addDebugLine("Quote.high does not grow consistently");
            return new BlockResultJava(screen.getLastQuote(), QUOTE_HIGH_NOT_GROWING_SCREEN_2);
        }
        EMAJava preLastEMA = CollectionUtilsJava.getFromEnd(screen_2_EMA13, 2);
        EMAJava lastEMA = CollectionUtilsJava.getFromEnd(screen_2_EMA13, 1);

        // both quotes below EMA - reject
        if (isQuoteBelowEMA(preLastQuote, preLastEMA.getValue()) && isQuoteBelowEMA(lastQuote, lastEMA.getValue())) {
            LogJava.addDebugLine("Both last two bars are below the EMA");
            LogJava.recordCode(QUOTES_BELOW_EMA_SCREEN_2, screen);
            return new BlockResultJava(screen.getLastQuote(), QUOTES_BELOW_EMA_SCREEN_2);
        }

        // both quotes above the EMA - reject
        if (isQuoteAboveEMA(preLastQuote, preLastEMA.getValue()) && isQuoteAboveEMA(lastQuote, lastEMA.getValue())) {
            LogJava.addDebugLine("Both last two bars are above the EMA");
            LogJava.recordCode(QUOTES_ABOVE_EMA_SCREEN_2, screen);
            return new BlockResultJava(screen.getLastQuote(), QUOTES_ABOVE_EMA_SCREEN_2);
        }

        // the penultimate one is below the EMA, the last one crosses or is above it - OK
        boolean crossingRule1 = isQuoteBelowEMA(preLastQuote, preLastEMA.getValue())
                && (isQuoteCrossedEMA(lastQuote, lastEMA.getValue()) || isQuoteAboveEMA(lastQuote, lastEMA.getValue()));

        // the penultimate one is below the EMA, the last one crosses - OK
        boolean crossingRule2 = isQuoteBelowEMA(preLastQuote, preLastEMA.getValue()) && isQuoteCrossedEMA(lastQuote, lastEMA.getValue());

        // penultimate and last cross the EMA - OK
        boolean crossingRule3 = isQuoteCrossedEMA(preLastQuote, preLastEMA.getValue()) && isQuoteCrossedEMA(lastQuote, lastEMA.getValue());

        // the penultimate one crosses the EMA, the last one is higher (it may be too late to enter the trade, we need to look at the chart) - OK
        boolean crossingRule4 = isQuoteCrossedEMA(preLastQuote, preLastEMA.getValue()) && isQuoteAboveEMA(lastQuote, lastEMA.getValue());

        boolean crossingOk = crossingRule1 || crossingRule2 || crossingRule3 || crossingRule4;
        if (!crossingOk) {
            LogJava.addDebugLine("The rule of EMA crossing is not fulfilled");
            LogJava.recordCode(CROSSING_RULE_VIOLATED_SCREEN_2, screen);
            return new BlockResultJava(screen.getLastQuote(), CROSSING_RULE_VIOLATED_SCREEN_2);
        }
        return new BlockResultJava(screen.getLastQuote(), OK);
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/two/Long_ScreenTwo_EMA_TwoBarsAscendingAndCrossing.kt
package com.kishlaly.ta.analyze.tasks.blocks.two

import com.kishlaly.ta.analyze.tasks.blocks.two.ScreenTwoBlock
import com.kishlaly.ta.model.*
import com.kishlaly.ta.model.indicators.EMA
import com.kishlaly.ta.model.indicators.Indicator
import com.kishlaly.ta.utils.CollectionUtils
import com.kishlaly.ta.utils.Log

/**
 * Quotes should cross EMA13 and should go up
 */
class Long_ScreenTwo_EMA_TwoBarsAscendingAndCrossing : ScreenTwoBlock {
    override fun check(screen: SymbolData): BlockResult {
        val screen_2_EMA13 = screen.indicator(Indicator.EMA13) as List<EMA>

        // prerequisite 1:
        // make sure first that the last TWO columns go up
        val ascendingBarHigh = screen.preLastQuote.high < screen.lastQuote.high
        if (!ascendingBarHigh) {
            Log.recordCode(BlockResultCode.QUOTE_HIGH_NOT_GROWING_SCREEN_2, screen)
            Log.addDebugLine("Quote.high does not grow consistently")
            return BlockResult(screen.lastQuote, BlockResultCode.QUOTE_HIGH_NOT_GROWING_SCREEN_2)
        }

        val preLastEMA = CollectionUtils.getFromEnd<EMA>(screen_2_EMA13, 2)
        val lastEMA = CollectionUtils.getFromEnd<EMA>(screen_2_EMA13, 1)

        // both quotes below EMA - reject
        if (screen.preLastQuote below preLastEMA && screen.lastQuote below lastEMA) {
            Log.addDebugLine("Both last two bars are below the EMA")
            Log.recordCode(BlockResultCode.QUOTES_BELOW_EMA_SCREEN_2, screen)
            return BlockResult(screen.lastQuote, BlockResultCode.QUOTES_BELOW_EMA_SCREEN_2)
        }

        // both quotes above the EMA - reject
        if (screen.preLastQuote above preLastEMA && screen.lastQuote above lastEMA) {
            Log.addDebugLine("Both last two bars are above the EMA")
            Log.recordCode(BlockResultCode.QUOTES_ABOVE_EMA_SCREEN_2, screen)
            return BlockResult(screen.lastQuote, BlockResultCode.QUOTES_ABOVE_EMA_SCREEN_2)
        }

        // the penultimate one is below the EMA, the last one crosses or is above it - OK
        val crossingRule1 =
            screen.preLastQuote below preLastEMA && (screen.lastQuote crosses lastEMA || screen.lastQuote above lastEMA)

        // the penultimate one is below the EMA, the last one crosses - OK
        val crossingRule2 = screen.preLastQuote below preLastEMA && screen.lastQuote crosses lastEMA

        // penultimate and last cross the EMA - OK
        val crossingRule3 = screen.preLastQuote crosses preLastEMA && screen.lastQuote crosses lastEMA

        // the penultimate one crosses the EMA, the last one is higher (it may be too late to enter the trade, we need to look at the chart) - OK
        val crossingRule4 = screen.preLastQuote crosses preLastEMA && screen.lastQuote above lastEMA

        val crossingOk = crossingRule1 || crossingRule2 || crossingRule3 || crossingRule4
        if (!crossingOk) {
            Log.addDebugLine("The rule of EMA crossing is not fulfilled")
            Log.recordCode(BlockResultCode.CROSSING_RULE_VIOLATED_SCREEN_2, screen)
            return BlockResult(screen.lastQuote, BlockResultCode.CROSSING_RULE_VIOLATED_SCREEN_2)
        }

        return BlockResult(screen.lastQuote, BlockResultCode.OK)
    }
}